dan_layer/consensus/src/hotstuff/block_transaction_executor.rs

```diff
@@ -0,0 +1,155 @@
+//   Copyright 2024 The Tari Project
+//   SPDX-License-Identifier: BSD-3-Clause
+
+use std::collections::HashSet;
+
+use log::info;
+use tari_dan_common_types::{Epoch, SubstateAddress};
+use tari_dan_engine::{
+    bootstrap_state,
+    state_store::{memory::MemoryStateStore, AtomicDb, StateWriter},
+};
+use tari_dan_storage::{consensus_models::{ExecutedTransaction, SubstateRecord}, StateStore, StateStoreReadTransaction, StorageError};
+use tari_engine_types::{substate::SubstateId, virtual_substate::{VirtualSubstate, VirtualSubstateId, VirtualSubstates}};
+use tari_epoch_manager::EpochManagerReader;
+use tari_transaction::{Instruction, Transaction};
+
+use crate::traits::{ConsensusSpec, TransactionExecutor};
+
+const LOG_TARGET: &str = "tari::dan::consensus::hotstuff::block_transaction_executor";
+
+// TODO: more refined errors
+#[derive(thiserror::Error, Debug)]
+pub enum BlockTransactionExecutorError {
+    #[error("Placeholder error")]
+    PlaceHolderError,
+    #[error("Execution thread failure: {0}")]
+    ExecutionThreadFailure(String),
+    #[error(transparent)]
+    StorageError(#[from] StorageError),
+    // TODO: remove this variant when we have a remote substate implementation
+    #[error("Remote substates are now allowed")]
+    RemoteSubstatesNotAllowed,
+}
+
+// TODO: we should keep a "proxy" hashmap (or memory storage) of updated states from previous transactions in the same
+// block,       so each consecutive transaction gets the most updated version for their inputs.
+//       If no previous tx wrote on a input, just get it from the regular state store
+#[derive(Debug, Clone)]
+pub struct BlockTransactionExecutor<TConsensusSpec: ConsensusSpec> {
```

Comment:
I think this should live in the validator node bin crate and be abstracted behind the TransactionExecutor trait. 
We should maintain a separation between the tari_engine and consensus.

The validator node bin can pass in the implementation.